core/src/main/java/jenkins/util/URLClassLoader2.java

```diff
@@ -0,0 +1,49 @@
+package jenkins.util;
+
+import jenkins.ClassLoaderReflectionToolkit;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import java.net.URL;
+import java.net.URLClassLoader;
+
+/**
+ * {@link URLClassLoader} with loosened visibility for use with {@link
+ * ClassLoaderReflectionToolkit}.
+ */
+@Restricted(NoExternalUse.class)
+public class URLClassLoader2 extends URLClassLoader implements JenkinsClassLoader {
+
+    static {
+        registerAsParallelCapable();
+    }
+
+    public URLClassLoader2(URL[] urls) {
+        super(urls);
+    }
+
+    public URLClassLoader2(URL[] urls, ClassLoader parent) {
+        super(urls, parent);
+    }
+
+    @Override
+    public void addURL(URL url) {
+        super.addURL(url);
+    }
```

Comment:
this is restricted - but do we generically want to add arbitrary URLs to the classloader?
would it be better to take a `File` or `Path` and then convert it?

```suggestion
    @Override
    public void addJar(File jar) {
        super.addURL(jar.toURI().toURL());
    }
```

Reply:
like `AntClassLoader.addPathComponent`

Reply:
As I wrote above, I do not intend to implement this RFE as part of this PR, though I consider it a reasonable idea.